### PR TITLE
[Fixes #4807] realize class + factory seperation

### LINKED ISSUE #4807: Dependency Injections are broken when using ember 2.12 and newer.
Originally filed as: https://github.com/emberjs/ember.js/issues/14941
Ember-Release (2.11.x): [Working Twiddle](https://ember-twiddle.com/e3a9f016848e28174059c52fe42944d2/75f5cfe06f78f0fee31f747ce66b5963e001d31e?openFiles=templates.application.hbs%2C)
Ember-Beta (2.12.x): [Broken Twiddle](https://ember-twiddle.com/e3a9f016848e28174059c52fe42944d2/7f6eb3760848bb88616fef26e854dc8fc3a055fe?openFiles=templates.application.hbs%2C)

It seems that using `owner.inject` targeting a `model` type is broken w/ ember-beta. Example:

```javascript
// app/initializers/i18n.js
export default {
  name: 'i18n',
  after: 'ember-i18n',
  initialize: function(app) {
    app.inject('model', 'i18n', 'service:i18n')
  }
};
```

Additionally, other ember types, such as component and controller, work as expected.

--------

@rwjblue helped me on slack to track this down a bit. We believe the cause of the problem lies with the implementation of `modelForFactory` ([store#modelForFactory](https://github.com/emberjs/data/blob/e9704ca5efa6e5bd8e0ca9e1b4014c9a33db386e/addon/-private/system/store.js#L2123-L2127)).

Prior to Ember 2.11, ember-data used `_lookupFactory` and internally that would do all of the dependency injections via an expensive `.extend()` process. One benefit of `factoryFor` is that it does not do this and thus saves perf. But the consequence is that when instantiating a class extracted from a "factory manager" those injections still need to be setup.

### PR DESCRIPTION
**Pull request by stefanpenner**
*note: this combined with embers factoryFor makes `MODEL_FACTORY_INJECTIONS` irrelevant.*

Fixes https://github.com/emberjs/data/issues/4807

### CODE REVIEW

File: addon/-private/system/store.js [hunk 1]
```diff
@@ -2101,9 +2107,13 @@ Store = Service.extend({
...
 
-      assert(`'${inspect(factory)}' does not appear to be an ember-data model`, (typeof factory._create === 'function') );
+      // interopt with the future
+      let klass = getOwner(this).factoryFor ? factory.class : factory;
+
+      assert(`'${inspect(klass)}' does not appear to be an ember-data model`, klass.isModel);
 
-      factory.modelName = factory.modelName || modelName;
+      // TODO: deprecate this
+      klass.modelName = klass.modelName || modelName;
```
Comment (chadhietala) [inline addon/-private/system/store.js]: Can you do `Ember.deprecate` instead of the TODO?

Comment (stefanpenner) [inline addon/-private/system/store.js]: We don't currently have a path forward yet. I would prefer to deprecate once this is the case.

Comment (stefanpenner) [inline addon/-private/system/store.js]: But will find a path forward for this soon!

Comment (stefanpenner) [inline addon/-private/system/store.js]: @chadhietala I'm doing a follow up PR that reduces ember-datas dependence on `modelName`. That will be a pre-cursor to deprecating it.

Comment (chadhietala) [inline addon/-private/system/store.js]: SG

Comment (stefanpenner) [inline addon/-private/system/store.js]: i've removed many more internal usages of `modelName`, there are some more low hanging fruit but addressing it fully will be some effort (although I hope we can do asap). 

Some notes: https://github.com/emberjs/data/issues/4814

**Pull request merged**